src/pudl/analysis/imputation/timeseries_cleaning.py

```diff
@@ -0,0 +1,1654 @@
+
+    Returns:
+        UTC times (tz-naive ``datetime64[ns]``).
+
+    Examples:
+        >>> s = pd.Series([pd.Timestamp(2020, 1, 1), pd.Timestamp(2020, 1, 1)])
+        >>> _local_to_utc(s, [-7, -6])
+        0   2020-01-01 07:00:00
+        1   2020-01-01 06:00:00
+        dtype: datetime64[ns]
+        >>> _local_to_utc(s, ['America/Denver', 'America/Chicago'])
+        0   2020-01-01 07:00:00
+        1   2020-01-01 06:00:00
+        dtype: datetime64[ns]
+    """
+    return local.groupby(tz, observed=True).transform(
+        lambda x: x.dt.tz_localize(
+            datetime.timezone(datetime.timedelta(hours=x.name))
+            if isinstance(x.name, int | float)
+            else x.name,
+            **kwargs,
+        ).dt.tz_convert(None)
+    )
+
+
+def _utc_to_local(utc: pd.Series, tz: Iterable) -> pd.Series:
+    """Convert UTC times to local.
+
+    Args:
+        utc: UTC times (tz-naive ``datetime64[ns]`` or ``datetime64[ns, UTC]``).
+        tz: For each time, a timezone (see :meth:`DatetimeIndex.tz_localize`)
+          or UTC offset in hours (``int`` or ``float``).
+
+    Returns:
+        Local times (tz-naive ``datetime64[ns]``).
+
+    Examples:
+        >>> s = pd.Series([pd.Timestamp(2020, 1, 1), pd.Timestamp(2020, 1, 1)])
+        >>> _utc_to_local(s, [-7, -6])
+        0   2019-12-31 17:00:00
+        1   2019-12-31 18:00:00
+        dtype: datetime64[ns]
+        >>> _utc_to_local(s, ['America/Denver', 'America/Chicago'])
+        0   2019-12-31 17:00:00
+        1   2019-12-31 18:00:00
+        dtype: datetime64[ns]
+    """
+    if utc.dt.tz is None:
+        utc = utc.dt.tz_localize("UTC")
+    return utc.groupby(tz, observed=True).transform(
+        lambda x: x.dt.tz_convert(
+            datetime.timezone(datetime.timedelta(hours=x.name))
+            if isinstance(x.name, int | float)
+            else x.name
+        ).dt.tz_localize(None)
+    )
+
+
+@pa.check_types
+def utc_dataframe_to_local(
+    input_df: DataFrame[UTCTimeseriesDataFrame],
+) -> DataFrame[LocalizedTimeseriesDataFrame]:
+    """Return DataFrame with localized ``datetime`` column."""
+    # Convert UTC datetimes to local datetimes (ignoring daylight savings)
+    input_df["utc_offset"] = input_df["timezone"].map(STANDARD_UTC_OFFSETS).astype(int)
+    input_df["datetime"] = _utc_to_local(
+        input_df["datetime_utc"],
+        input_df["utc_offset"],
+    )
+
+    # List timezone by year for each respondent by the datetime
+    input_df["year"] = input_df["datetime"].dt.year
+    return input_df
+
+
+@pa.check_types
+def local_dataframe_to_utc(
+    input_df: DataFrame[LocalizedTimeseriesDataFrame],
+) -> DataFrame[UTCTimeseriesDataFrame]:
+    """Return DataFrame with localized ``datetime`` converted back to ``datetime_utc``."""
+    # Convert local times to UTC
+    input_df["datetime_utc"] = _local_to_utc(
+        input_df["datetime"], input_df["utc_offset"]
+    )
+    return input_df
+
+
+@dataclass
+class FlaggedTimeseries:
```

Comment:
TUrned this into just a data type that gets passed around, because the state management of the old class was really confusing and felt error prone to me.

Reply:
Otherwise haven't modified the methods below